anthos-bm-gcp-terraform/modules/init/main.tf

```diff
@@ -0,0 +1,82 @@
+locals {
+  ssh_pub_key_template_file = "${path.module}/../../resources/ssh-keys.tpl"
+  ssh_pub_key_file          = format(var.pub_key_path_template, var.hostname)
+  ssh_private_key_file      = format(var.priv_key_path_template, var.hostname)
+  cluster_yaml_file_name    = trimprefix(basename(var.cluster_yaml_path), ".")
+}
+
+resource "tls_private_key" "ssh_key_pair" {
+  algorithm = "RSA"
+  rsa_bits  = 4096
+}
+
+resource "local_file" "temp_ssh_pub_key_file" {
+  filename        = local.ssh_pub_key_file
+  file_permission = "0600"
+  content = templatefile(
+    local.ssh_pub_key_template_file,
+    { ssh_key = chomp(tls_private_key.ssh_key_pair.public_key_openssh) }
+  )
+}
+
+resource "local_file" "temp_ssh_priv_key_file" {
+  filename        = local.ssh_private_key_file
+  file_permission = "0600"
+  content         = chomp(tls_private_key.ssh_key_pair.private_key_pem)
+}
+
+module "gcloud_add_ssh_key_metadata" {
+  source                   = "terraform-google-modules/gcloud/google"
+  version                  = "2.0.3"
+  platform                 = "linux"
+  service_account_key_file = var.credentials_file
+  module_depends_on = [
+    local_file.temp_ssh_pub_key_file,
+    local_file.temp_ssh_priv_key_file
+  ]
+
+  create_cmd_entrypoint = "gcloud"
+  create_cmd_body       = <<EOT
+    compute instances add-metadata ${var.hostname}  \
+    --project ${var.project_id}                                    \
+    --zone ${var.zone}                                             \
+    --metadata-from-file ssh-keys=${local.ssh_pub_key_file}
+  EOT
+}
+
+resource "null_resource" "exec_init_script" {
+  depends_on = [module.gcloud_add_ssh_key_metadata]
+  connection {
+    type        = "ssh"
+    user        = "root"
+    host        = var.publicIp
+    private_key = chomp(tls_private_key.ssh_key_pair.private_key_pem)
+  }
+
+  provisioner "file" {
+    source      = var.cluster_yaml_path
+    destination = "/root/${local.cluster_yaml_file_name}"
+  }
+
+  provisioner "file" {
+    source      = var.init_script
+    destination = "/root/init.sh"
+  }
+
+  provisioner "remote-exec" {
+    inline = [
+      "chmod 0100 /root/init.sh",
+      "chmod 0100 /root/env.sh",
+      "chmod 0600 /root/${local.cluster_yaml_file_name}"
+    ]
+  }
+
+  provisioner "local-exec" {
+    command = <<EOT
+      ssh -o 'StrictHostKeyChecking no' \
+      -i ${local.ssh_private_key_file} \
+      root@${var.publicIp} \
+      'echo "${var.hostnames}" > hostnames && echo "${var.internalIps}" > internalIps && nohup /root/init.sh ${var.init_script_args} > /root/init.log 2>&1 &'
```

Comment:
What is init.sh fails? How will we know and what action will be taken?

Reply:
One way is to use return codes and use them for either message or retry.

Reply:
Yeah. I will work on that. Will have to make the script more `pure` so retries don't break if half the steps succeeded in the previous run.

Reply:
I have added status messages per step in the initialization script which is based on the exit status. In addition I have also added a `preflights.sh` script that is run against the log file _(for the init script)_ to check if all the initialization steps passed, before executing `bmctl`

See https://github.com/GoogleCloudPlatform/anthos-samples/pull/11/commits/1dec940a14616e7a1935bdfaa64f4da0cd108e2b and https://github.com/GoogleCloudPlatform/anthos-samples/pull/11/commits/895cbceb3b264922422b2014a77cfe27d1ad1b05

I have made this asynchronous since it might be time consuming to wait until each host completes initialization before moving to the next.

Currently the `preflight.sh` check only validates the initialization of the admin host. I am working on seeing if I cna improve this script to check the logs of other hosts over SSH too